### PR TITLE
[CFM-311] contact multiple members

### PR DESCRIPTION
**Pull request by anvmn**
https://issuetracker.amplexor.com/jira/browse/CFM-1259
https://issuetracker.amplexor.com/jira/browse/CFM-1260

--
https://issuetracker.amplexor.com/jira/browse/CFM-311

### CODE REVIEW

**Comment by anvmn** [comment]
@ordavidil 
Travis is not activated on push. I verified that coder passes.

File: project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss [hunk 1]
```diff
@@ -92,3 +92,24 @@
...
+  .views-field-rendered-entity,
+  .views-field-views-bulk-operations {
+    display: inline-block;
+  }
+
+  .views-field-views-bulk-operations {
+    vertical-align: text-bottom;
+    margin-bottom: 12px;
+  }
+
```
Comment (ordavidil) [inline project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss]: remove empty line.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user_admin/c4m_user_admin.install [hunk 2]
```diff
@@ -14,3 +14,10 @@ function c4m_user_admin_install() {
   $views_status['admin_views_user'] = TRUE;
   variable_set('views_defaults', $views_status);
 }
+
+/**
+ * Enables JQuery Migration plugin.
+ */
+function c4m_user_admin_update_7000() {
+  variable_set('jquery_update_jquery_migrate_enable', TRUE);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user_admin/c4m_user_admin.install]: there is no need to use updates yes since it is still not on production, move this to `capacity4more_setup_set_variables` (you can find it in `capacity4more.profile` file)

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 3]
```diff
@@ -1286,3 +1299,51 @@ function c4m_user_ckeditor_settings_alter(&$settings, $conf) {
...
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Updates value of submit button.
+ */
+function c4m_user_form_views_form_c4m_overview_og_members_page_alter(&$form, &$form_state, $form_id) {
+  if (isset($form['actions'])) {
+    $form['actions']['submit']['#value'] = 'Send';
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: => `t('Send')`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 4]
```diff
@@ -1286,3 +1299,51 @@ function c4m_user_ckeditor_settings_alter(&$settings, $conf) {
...
+
+/**
+ * Implements hook_action_info_alter().
+ *
+ * Removes 'contact multiple' action, if user is not admin / group owner.
+ */
+function c4m_user_action_info_alter(&$actions) {
+  global $user;
+
+  if ($user->uid == 1) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: I believe the meaning was the GA and not the SA, hence you probably want to use the `og_user_access()` here

File: project/profiles/capacity4more/modules/c4m/features_og/c4m_features_og_members/c4m_features_og_members.context.inc [hunk 5]
```diff
@@ -1,5 +1,4 @@
 <?php
-
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features_og/c4m_features_og_members/c4m_features_og_members.context.inc]: revert this

File: project/profiles/capacity4more/modules/c4m/features_og/c4m_features_og_members/c4m_features_og_members.features.inc [hunk 6]
```diff
@@ -1,5 +1,4 @@
 <?php
-
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features_og/c4m_features_og_members/c4m_features_og_members.features.inc]: revert this

File: project/profiles/capacity4more/modules/c4m/features_og/c4m_features_og_members/c4m_features_og_members.views_default.inc [hunk 7]
```diff
@@ -1,5 +1,4 @@
 <?php
-
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/features_og/c4m_features_og_members/c4m_features_og_members.views_default.inc]: revert this

File: project/profiles/capacity4more/capacity4more.profile [hunk 8]
```diff
@@ -89,8 +89,10 @@ function capacity4more_setup_set_variables(&$install_state) {
     'jquery_update_jquery_admin_version' => '2.1',
     'page_manager_node_view_disabled' => FALSE,
     'page_manager_term_view_disabled' => FALSE,
+    'jquery_update_jquery_migrate_enable' => TRUE,
 
-    // RESTful.
```
Comment (ordavidil) [inline project/profiles/capacity4more/capacity4more.profile]: revert this change

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 9]
```diff
@@ -1286,3 +1299,44 @@ function c4m_user_ckeditor_settings_alter(&$settings, $conf) {
...
+  }
+}
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Updates value of submit button.
+ */
+function c4m_user_form_views_form_c4m_overview_og_members_page_alter(&$form, &$form_state, $form_id) {
+  if (isset($form['actions'])) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: remove this condition.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+  );
+}
+
+/**
+ * Action function for contact_multiple_form.
+ *
+ * Activated for each selected account.
+ * Sends email to account.
+ */
+function c4m_user_contact_multiple_action(&$account, array $context) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: @anvmn 
remove the `&`, since we don't really change the account.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 11]
```diff
@@ -0,0 +1,164 @@
...
+ */
+function c4m_user_contact_multiple_action_form(array $context, $form_state) {
+  global $user;
+  $to_list = '';
+
+  // If 'Select all items on all pages' is selected.
+  if ($form_state['select_all_pages']) {
+    $group = c4m_og_current_group();
+    // Get ids of all subscribers of current group.
+    $ids = og_get_group_members_properties($group, array(), 'members', 'node');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: => `$uids`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 12]
```diff
@@ -0,0 +1,164 @@
...
+  }
+  else {
+    // Get selected items on current page.
+    $ids = $form_state['selection'];
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: => `$uids`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 13]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  // Create a names list of notified subscribers.
+  foreach ($ids as $account_id) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: =>
```PHP
foreach ($uids as $uid) {
```

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 14]
```diff
@@ -0,0 +1,164 @@
...
+    // Make sure that id is valid, and GO/GA didn't select himself.
+    if ($account_id != 0 && $user->uid != $account_id) {
+      $account = user_load($account_id);
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: @anvmn 
we should use `user_load_multiple()` which will hit the DB only once, instead of hitting in for each user.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 15]
```diff
@@ -0,0 +1,164 @@
...
+      $to_list .= c4m_user_name($account) . ', ';
+    }
+  }
+  $to_list = rtrim($to_list, ", ");
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: please use:
```PHP
implode(',', $users_names);
```

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 16]
```diff
@@ -0,0 +1,164 @@
...
+    ),
+  );
+}
+
+/**
+ * Create 'Contact Multiple' form.
+ */
+function c4m_user_contact_multiple_action_form(array $context, $form_state) {
+  global $user;
+  $to_list = '';
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: =>
```PHP
$users_names = array();
```

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 17]
```diff
@@ -1286,3 +1299,46 @@ function c4m_user_ckeditor_settings_alter(&$settings, $conf) {
...
+  unset($actions['c4m_user_contact_multiple_action']);
+}
+
+/**
+ * Implements hook_module_implements_alter().
+ *
+ * Ensures c4m_user form_alter runs after VBO.
+ */
+function c4m_user_module_implements_alter(&$implementations, $hook) {
+  if ($hook == 'form_alter') {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: @anvmn 
please use return early over here.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 18]
```diff
@@ -0,0 +1,164 @@
+<?php
+
+/**
+ * @file
+ * Callback & forms related to the user contact multiple.
+ */
+
+/**
+ * A custom action to contact a multiple users with a mail action.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: @anvmn 
please add
```PHP
Implements hook_action_info().
```

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 19]
```diff
@@ -1286,3 +1299,46 @@ function c4m_user_ckeditor_settings_alter(&$settings, $conf) {
     unset($settings['loadPlugins']['media']);
   }
 }
+
+/**
+ * Implements hook_action_info_alter().
+ *
+ * Removes 'contact multiple' action, if user is not admin / group owner.
+ */
+function c4m_user_action_info_alter(&$actions) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: Remove this function and add the logic to `function c4m_user_contact_multiple_action_info()` on line 14

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 20]
```diff
@@ -0,0 +1,173 @@
...
+  $group = c4m_og_current_group();
+  // If not within OG context, don't declare action.
+  if (!$group) {
+    return array();
+  }
+  // If not Site Admin, Group Admin or Group Owner, don't declare action.
+  if (!_c4m_features_og_members_is_power_user($group)) {
+    return array();
+  }
+  // Site Admin, Group Admin or Group Owner => declare action.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: 1. Move this comment to the docs of the function.
2. Leave a line spacing above the `return` command.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 21]
```diff
@@ -0,0 +1,173 @@
...
+      'type' => 'user',
+      'label' => t('Contact'),
+      'configurable' => TRUE,
+      'triggers' => array('any'),
+    ),
+  );
+}
+
+/**
+ * Create 'Contact Multiple' form.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: It looks like this is an implementation of _hook_form_, we should declare this on the docs (see other hooks).

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 22]
```diff
@@ -0,0 +1,173 @@
...
+ */
+function c4m_user_contact_multiple_action_form(array $context, $form_state) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: It doesn't look like the right signature of a _hook_form_

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc (hunk 21 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: Action form function.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc (hunk 22 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: This is VBO action definition standard.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc (hunk 21 above)
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: so please add `Implements hook_action_info().`

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: ok, it is not this hook, but it is looks like a _hook_form_

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: check this one:
https://api.drupal.org/api/drupal/modules!node!node.api.php/function/hook_form/7.x

File: project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss [hunk 23]
```diff
@@ -92,3 +92,23 @@
 .panel-title {
   text-transform: uppercase;
 }
+
+#views-form-c4m-overview-og-members-page {
+
+  .vbo-fieldset-select-all {
+    width: 250px;
+    text-align: left;
+    padding-top: 20px !important;
```
Comment (ordavidil) [inline project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss]: @anvmn 
why do we need the `!important` over here ?

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 24]
```diff
@@ -0,0 +1,176 @@
...
+    $uids = $form_state['selection'];
+  }
+
+  $accounts = user_load_multiple($uids);
+  $users_name = array();
+
+  // Create an array of users names for notified subscribers.
+  foreach ($accounts as $account) {
+    // Make sure that id is valid, and GO/GA didn't select himself.
+    if ($account->uid != 0 && $user->uid != $account->uid) {
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: @anvmn 
Is it possible to select an anonymous ?

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 25]
```diff
@@ -0,0 +1,176 @@
...
+ * Create 'Contact Multiple' form.
+ */
+function c4m_user_contact_multiple_action_form(array $context, $form_state) {
+  global $user;
+
+  // If 'Select all items on all pages' is selected.
+  if ($form_state['select_all_pages']) {
+    $group = c4m_og_current_group();
+    // Get ids of all active subscribers of current group.
+    $uids = og_get_group_members_properties($group, array(), 'members__' . OG_STATE_ACTIVE, 'node');
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: @amitaibu 
Is this fine to use this getter over here ?

File: project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss (hunk 23 above)
Comment (anvmn) [inline project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss]: @ordavidil 
Originally, it was supposed to override another class with !important there.
But I rechecked, and it's not really required.
I'll remove it.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc [hunk 26]
```diff
@@ -0,0 +1,176 @@
...
+
+/**
+ * Action form function.
+ *
+ * Create 'Contact Multiple' form.
+ */
+function c4m_user_contact_multiple_action_form(array $context, $form_state) {
+  global $user;
+
+  // If 'Select all items on all pages' is selected.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: @anvmn 
We should select members, and not items, please fix that also on the shown label on the page.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc (hunk 24 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: @ordavidil 
No, it's imposible.
uid != 0 check if to make sure that og_get_group_members_properties() doesn't return some account with invalid uid.
Should I remove it?

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc (hunk 26 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/actions/c4m_user_contact_multiple.action.inc]: @ordavidil 
I didn't find an option to change this label at form.
VBO adds ckeckbox fields at  $form['select_all']['this_page'],  $form['select_all']['all_pages'], but when viewing the form, these fields are not available.

File: project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss [hunk 27]
```diff
@@ -92,3 +92,23 @@
 .panel-title {
   text-transform: uppercase;
 }
+
+#views-form-c4m-overview-og-members-page {
+
+  .vbo-fieldset-select-all {
+    width: 250px;
```
Comment (ordavidil) [inline project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss]: Did you check this one on mobile resolution ?

Comment (anvmn) [inline project/profiles/capacity4more/themes/c4m/kapablo/build/src/stylesheets/components/_forms.scss]: @ordavidil 
Yes. I added a comment with mobile view.

**Comment by anvmn** [comment]
selection_107

**Comment by ordavidil** [comment]
@anvmn 
The screenshots still shows __Select all items__...

**Comment by anvmn** [comment]
@ordavidil - After clean cache:

selection_108

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 28]
```diff
@@ -1286,3 +1299,36 @@ function c4m_user_ckeditor_settings_alter(&$settings, $conf) {
...
+
+/**
+ * Implements hook_form_FORM_ID_alter().
+ *
+ * Updates value of submit button.
+ */
+function c4m_user_form_views_form_c4m_overview_og_members_page_alter(&$form, &$form_state, $form_id) {
+  $form['actions']['submit']['#value'] = t('Send');
+  if (isset($form['select_all_markup']['#markup'])) {
+    // VBO module doesn't provide an option to change 'Select all' labels.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: `labels` => `label`

Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: also replace the `.` with `,` at the and of the line

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 29]
```diff
@@ -1286,3 +1299,36 @@ function c4m_user_ckeditor_settings_alter(&$settings, $conf) {
...
+    // Only option is to change at markup.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: => `hence the only option is to replace the text on the markup.`

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module (hunk 28 above)
Comment (anvmn) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: @ordavidil 
It's labels, as in case there're more users than defined to be displayed on page, there're second label: Select all members on all pages.

File: project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module [hunk 30]
```diff
@@ -813,9 +819,11 @@ function c4m_user_contact_form_submit($form, $form_state) {
  * @param string $body
  *   Email message body.
  * @param object $language
- *   Addressee preferred language.
+ *   Preferred language for addressee user.
+ * @param bool $silent
+ *   Wither to present status message or not.
```
Comment (ordavidil) [inline project/profiles/capacity4more/modules/c4m/user/c4m_user/c4m_user.module]: => `Whether...`

**Comment by jover** [comment]
CFM-1200 is already merged into develop, so this PR should be merged into develop instead of CFM-1200

**Comment by ordavidil** [comment]
@anvmn 
please close this one and open it comparing to develop.

**Comment by anvmn** [comment]
Closing in favor of #1107
